src/config.ts

```diff
@@ -0,0 +1,246 @@
+"use strict";
+
+import { FirebaseConfig } from "./firebaseConfig";
+
+const _ = require("lodash");
+const clc = require("cli-color");
+const cjson = require("cjson");
+const fs = require("fs-extra");
+const path = require("path");
+
+const detectProjectRoot = require("./detectProjectRoot").detectProjectRoot;
+const { FirebaseError } = require("./error");
+const fsutils = require("./fsutils");
+const loadCJSON = require("./loadCJSON");
+const parseBoltRules = require("./parseBoltRules");
+const { promptOnce } = require("./prompt");
+const { resolveProjectPath } = require("./projectPath");
+const utils = require("./utils");
+
+export class Config {
+  static FILENAME = "firebase.json";
+  static MATERIALIZE_TARGETS = [
+    "database",
+    "emulators",
+    "firestore",
+    "functions",
+    "hosting",
+    "storage",
+    "remoteconfig",
+  ];
+
+  public options: any;
+  public projectDir: string;
+  public data: any = {};
+  public defaults: any = {};
+  public notes: any = {};
+
+  private _src: any;
+
+  constructor(src: any, options: any) {
+    this.options = options || {};
+    this.projectDir = options.projectDir || detectProjectRoot(options);
+    this._src = src;
+
+    if (this._src.firebase) {
+      this.defaults.project = this._src.firebase;
+      utils.logWarning(
+        clc.bold('"firebase"') +
+          " key in firebase.json is deprecated. Run " +
+          clc.bold("firebase use --add") +
+          " instead"
+      );
+    }
+
+    if (_.has(this._src, "rules")) {
+      _.set(this._src, "database.rules", this._src.rules);
+    }
+
+    Config.MATERIALIZE_TARGETS.forEach((target) => {
+      if (_.get(this._src, target)) {
+        _.set(this.data, target, this._materialize(target));
+      }
+    });
+
+    // auto-detect functions from package.json in directory
+    if (
+      this.projectDir &&
+      !this.get("functions.source") &&
+      fsutils.fileExistsSync(this.path("functions/package.json"))
+    ) {
+      this.set("functions.source", "functions");
+    }
+  }
+
+  _hasDeepKey(obj: any, key: string) {
```

Comment:
Why are we still using underscore methods instead of `private` annotations?

Reply:
Just because my intent was to change this file as little as possible in this PR. I found at least one case in the codebase where another file was accessing what should be a "private" method/property of Config.